nixos/modules/programs/dconf.nix

```diff
@@ -3,12 +3,94 @@
 let
   cfg = config.programs.dconf;
 
+  # Compile keyfiles to dconf DB
+  compileDconfDb = dir: pkgs.runCommand "dconf-db"
+    {
+      nativeBuildInputs = [ (lib.getBin pkgs.dconf) ];
+    } "dconf compile $out ${dir}";
+
+  # Generate dconf DB from attrset and keyfiles
+  mkDconfDb = val: compileDconfDb (pkgs.symlinkJoin {
+    name = "nixos-generated-dconf-keyfiles";
+    paths = [
+      (pkgs.writeTextDir "nixos-generated-dconf-keyfiles" (lib.generators.toDconfINI val.settings))
+      val.keyfiles
+    ];
+  });
+
   # Generate dconf profile
   mkDconfProfile = name: value:
-    pkgs.runCommand "dconf-profile" { } ''
-      mkdir -p $out/etc/dconf/profile/
-      cp ${value} $out/etc/dconf/profile/${name}
-    '';
+    if lib.isDerivation value || lib.isPath value then
+      pkgs.runCommand "dconf-profile" { } ''
+        mkdir -p $out/etc/dconf/profile/
+        cp ${value} $out/etc/dconf/profile/${name}
+      ''
+    else
+      pkgs.writeTextDir "etc/dconf/profile/${name}" (
+        lib.concatMapStrings (x: "${x}\n") ((
+          lib.optional value.enableUserDb "user-db:user"
+        ) ++ (
+          map
+            (value:
+              let
+                db = if lib.isAttrs value && !lib.isDerivation value then mkDconfDb value else value;
+              in
+              "file-db:${db}")
+            value.databases
+        ))
+      );
+
+  dconfProfile = with lib.types; submodule {
+    options = {
+      enableUserDb = lib.mkOption {
+        type = bool;
+        default = true;
+        description = lib.mdDoc "Add `user-db:user` at the beginning of the profile.";
+      };
+
+      databases = lib.mkOption {
+        type = with lib.types; listOf (oneOf [
+          path
+          package
```

Comment:
The path is writen to the profile as a file-db. Do I need to make sure it's a file? I thought the user should ensure that it's a valid dconf database.

Reply:
Yeah, ideally, the API should prevent errors as soon as possible, when practical. For example, if user passes `dbPackage` that really contains a directory with `db.conf`, it would fail. Though, I am not sure if it is possible to distinguish that. Maybe it is actually not practical :woman_shrugging:

Reply:
Maybe we can write a key to the database.

Reply:
Not sure what you mean by that.

Reply:
My fault. `dconf write` can't write to a specfic database file. I thought we can only check if it's a file. Should I add that?

Reply:
Right, `dconf write` only touches `user-db`. I am still not sure how writing is relevant to linking databases from profile.

File check would be nice but it is probably not possible to do it at module system type level (without IFD).

Maybe replacing `pkgs.writeTextDir "etc/dconf/profile/${name}"` with `writeTextFile` and checking in its `checkPhase` would work, though.

Reply:
How about checking and copying in runCommand?

Reply:
Now it's checked in runCommand to be a valid dconf database.

Reply:
Thanks. That should work.